### ISSUE TITLE
optim_lbfgs gives errors

### ISSUE THREAD

**Issue by mchikina** [mlverse/torch#1032]
trying the lbfgs optimizer results in this error. version 0.10.
Error in closure() : argument "closure" is missing, with no default

**Comment by dfalbel** [comment]
Hi @mchikina,

Since LBFGS is a second order optimizer, you need to pass a closure to the `step` for it to work correctly, eg:

```
x <- torch_tensor(1, requires_grad = TRUE)
opt <- optim_lbfgs(x)
opt$zero_grad()
f <- function() {
  y <- 2 * x
}
opt$step(f)```

The closure is any function that returns a scalar value, in general it's a closure that computes the loss value. Internally, when calling `step()`, `backward()` will be called on its result and parameters will be updated accordingly.

**Comment by mchikina** [comment]
I think I still don't understand something. I have code that runs but the weights are not updated. Here is a code snippet running a simple MLP on a toy problem. Could you advise how I can fix this?



###
#Model
###

mlp <- nn_module(
  initialize = function(input, H=10, output=1) {
    self$linear1=nn_linear(input,H)
    self$linear2=nn_linear(H,output) 
  },
  forward = function(x) {
    x %>% 
      self$linear1() %>%
      nnf_relu() %>%
      self$linear2() %>% torch_squeeze()
  }
)

model=mlp(input)

optimizer <- optim_lbfgs(model$parameters, lr = 1, history_size = 50)
losses=double()
lossesTest=double()
for (epoch in 1:200) {
  for (b in enumerate(train_dl)) {
    clo=function(){
      output <- model(b[[1]])
    loss=nnf_binary_cross_entropy_with_logits(output, (b[[2]]-1))
    return(loss)
    }
    
    # apply weight updates
    optimizer$step(clo)
  }
}

**Comment by dfalbel** [comment]
I'm sorry @mchikina , my bad, my answer above is incorrect. You have to manually call backward inside the closure function, so eg, `clo` should be:

```
clo=function(){
 output <- model(b[[1]])
 loss=nnf_binary_cross_entropy_with_logits(output, (b[[2]]-1))
 loss$backward()
 return(loss)
}
```

**Comment by mchikina** [comment]
Thank you so much for all your help. That was indeed the difference. However, now the optimizer runs but the loss blows up. If I switch between adam and lbfgs  on exactly the same problem adam converges nicely and lbfgs rapidly increases the loss. Do you have any suggestions? Here is the code I am using:
  for (b in enumerate(train_dl)) {
  optimizer$zero_grad()   
    if(use.lbfgs){
      clo=function(){
        output <- model(b[[1]])
        loss=nnf_binary_cross_entropy_with_logits(output, (b[[2]]-1))
        loss$backward()
        print(as.numeric(loss))
        return(loss)
      }
      optimizer$step(clo)
    }
    else{
      output <- model(b[[1]])
      loss=nnf_binary_cross_entropy_with_logits(output, (b[[2]]-1))
      loss$backward()
      optimizer$step()
    }
  }

**Comment by skeydan** [comment]
Maybe using line search helps - here's  a short example:

https://skeydan.github.io/Deep-Learning-and-Scientific-Computing-with-R-torch/optim_2.html

**Comment by dfalbel** [comment]
Docs improved in #1035 !
@mchikina I'm closing this issue as resolved. Please reopen or create a new issue if you need additional help.
